src/librustdoc/html/format.rs

```diff
@@ -1085,32 +1085,43 @@ impl Function<'_> {
 }
 
 impl clean::Visibility {
-    crate fn print_with_space<'tcx>(self, tcx: TyCtxt<'tcx>) -> impl fmt::Display + 'tcx {
+    crate fn print_with_space<'tcx>(
+        self,
+        tcx: TyCtxt<'tcx>,
+        item_did: DefId,
+    ) -> impl fmt::Display + 'tcx {
         use rustc_span::symbol::kw;
 
         display_fn(move |f| match self {
             clean::Public => f.write_str("pub "),
             clean::Inherited => Ok(()),
-            clean::Visibility::Restricted(did) if did.index == CRATE_DEF_INDEX => {
-                write!(f, "pub(crate) ")
-            }
-            clean::Visibility::Restricted(did) => {
-                f.write_str("pub(")?;
-                let path = tcx.def_path(did);
-                debug!("path={:?}", path);
-                let first_name =
-                    path.data[0].data.get_opt_name().expect("modules are always named");
-                if path.data.len() != 1 || (first_name != kw::SelfLower && first_name != kw::Super)
-                {
-                    f.write_str("in ")?;
-                }
-                // modified from `resolved_path()` to work with `DefPathData`
-                let last_name = path.data.last().unwrap().data.get_opt_name().unwrap();
-                for seg in &path.data[..path.data.len() - 1] {
-                    write!(f, "{}::", seg.data.get_opt_name().unwrap())?;
+
+            clean::Visibility::Restricted(vis_did) => {
+                if find_closest_parent_module(tcx, item_did) == Some(vis_did) {
+                    // `pub(in foo)` where `foo` is the parent module
+                    // is the same as no visibility modifier
+                    Ok(())
```

Comment:
Hmm, I'm not sure how this will behave for modules - can you add a test case?

Reply:
What do you mean?

Reply:
If you have `mod x;`, `x` is public, but the parent will be the current module, not `x`. So it's actually public in the current module, not in `x`. I'm not sure how your code will behave in that case.

Reply:
@jyn514 I don't think rustdoc even renders visibility for `mod`s though.

Reply:
Hmm, you're right: https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/passes/calculate_doc_coverage/index.html

It might want to in the future though - can you leave a `FIXME`? r=me with that done.

Reply:
Done.